### ISSUE TITLE
Errors are not shown if fields were autocompleted by the browser

### ISSUE THREAD

**Issue by youroff** [kaliber5/ember-bootstrap#611]
I use ember-bootstrap with ember-changeset. When I add errors programmatically after api call (server side errors), they don't appear on the page. This happens if the field was prefilled (autocompleted) by the browser, so no `focusout` or any other input event happened to trigger `showOwnValidation`.

I haven't looked into the implementation thoroughly yet, but it seems like the decision on wether errors should be displayed is made not only based on presence of anything in error property, but also on some preconditions which makes for this issue. I assume this was made to make the successful validation case work, but should they really be connected?

Is this an actual issue? is there a non ugly workaround for that? If that's an issue, I probably could invest some time into fixing that. Thanks!

Output from `ember bootstrap:info` (available since 1.0.0-alpha.9):
```
Npm packages:
ember-bootstrap: ^1.2.2 -> 1.2.2
ember-cli: ~3.1.4 -> 3.1.4
bootstrap: ^4.1.1 -> 4.1.1
bootstrap-sass: n/a
ember-cli-sass: ^7.1.7 -> 7.1.7
ember-cli-less: n/a
Bower packages:
bootstrap: n/a
bootstrap-sass: n/a
ember-bootstrap configuration:
bootstrapVersion: 4
importBootstrapFont: false
importBootstrapCSS: false
```

**Comment by simonihmig** [comment]
Yes, this is indeed intended. Validations are only shown on `focusout` or when submitting the form. Simple reason: you have an form element with a "required"  validation, but you don't want to show the error state when a blank form is initially shown. So the form needs some user interaction, that signals that the given form elements value has been entered (which is by default a `focusout` or `submit` event).

There was another request for programmatically controlling the display of errors: #364. 

I believe your issue is a duplication of that, so closing this for now. Feel free to reopen if I misunderstood your use case!

**Comment by youroff** [comment]
That issue is similar but subtly different, I suppose.

I think it should be the other way around: validation (of particular field) should happen on focusout or submit, but errors should always be displayed if present. Cause if user (of the library) doesn't pay attention to such details, they might end up with user of the app unable to submit and seeing no errors whatsoever.

If errors showed up after submit, that would be sound and totally worked, but in my case it's not happening. Is there something else that tells bs-form to clear out related flags?

**Comment by simonihmig** [comment]
Maybe it would help if I had more experience with `ember-changeset`, which I unfortunately don't have...

> validation (of particular field) should happen on focusout or submit, but errors should always be displayed if present.

In `ember-cp-validations` there is no distinction between those (validation and having errors). As soon as you have an empty property which has a non-empty validation defined, this property has an error. Without you having to explicitly call `validate()` or so. Hence the logic I explained above...

Would setting `showAllValidations` to true (although being private) on the form component solve your use case?

If not, I think it would be helpful to have a minimal repo that demonstrates your use case. Would that be possible?

**Comment by simonihmig** [comment]
> What would fix that, if it showed all after submit, that would allow to show server-side errors, which is exactly my use case.

Hm, but that should be happing here: https://github.com/kaliber5/ember-bootstrap/blob/master/addon/components/base/bs-form.js#L233-L241.

Maybe the implementation in `ember-bootstrap-changeset-validations` (you are using this addon, right?) is not correct though, as it seems to not care about async validations (which I guess is the case for server-side errors): https://github.com/kaliber5/ember-bootstrap-changeset-validations/blob/master/addon/components/bs-form.js#L20

**Comment by youroff** [comment]
Yeah, probably... I'll start by integrating it myself then.

**Comment by youroff** [comment]
Oh wait, seems like it's not happening on the happy path (when client side validations passed), `showAllValidations` stays `false`. Also what would be the situation regarding showing errors, if there's no client-side validation, but only server-side, that would just populate model with errors on failure? Would it make sense to just set `showAllValidations` true on submit attempt unconditionally?

**Comment by jelhan** [comment]
If I remember correctly validation errors are shown if

- an event listed in `showValidationOn` property of FormElement has been fired or
- if from has been submitted.

Implementation wise the first one sets `showOwnValidation` property of `FormElement` to true. The later one `showAllValidations` property of `Form`. Visibility of validations is controlled by `showValidation` computed property of `FormElement`.

`showValidationOn` defaults to `focusOut` event. I'm not quite sure if autocompletion of browser is firing any event but for sure it's not firing `focusOut` event. Therefore it does not change validation visibility. From UI/UX perspective I'm also not sure if it should. But for sure from a UI/UX perspective it's not a good idea to disable submit button if form is invalid, cause in that case user is not able to trigger an event that shows validation errors.
